drivers/sensor/vcnl36825t/vcnl36825t.c

```diff
@@ -0,0 +1,478 @@
+/*
+ * Copyright (c) 2023 Juliane Schulze, deveritec GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT vishay_vcnl36825t
+
+#include "vcnl36825t.h"
+
+#include <stdlib.h>
+
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/util.h>
+
+LOG_MODULE_REGISTER(VCNL36825T, CONFIG_SENSOR_LOG_LEVEL);
+
+static int vcnl36825t_read(const struct i2c_dt_spec *spec, uint8_t reg_addr, uint16_t *value)
+{
+	uint8_t rx_buf[2];
+	int rc;
+
+	rc = i2c_write_read_dt(spec, &reg_addr, sizeof(reg_addr), rx_buf, sizeof(rx_buf));
+	if (rc < 0) {
+		return rc;
+	}
+
+	*value = sys_get_le16(rx_buf);
+
+	return 0;
+}
+
+static int vcnl36825t_write(const struct i2c_dt_spec *spec, uint8_t reg_addr, uint16_t value)
+{
+	uint8_t tx_buf[3] = {reg_addr};
+
+	sys_put_le16(value, &tx_buf[1]);
+	return i2c_write_dt(spec, tx_buf, sizeof(tx_buf));
+}
+
+static int vcnl36825t_update(const struct i2c_dt_spec *spec, uint8_t reg_addr, uint16_t mask,
+			     uint16_t value)
+{
+	int rc;
+	uint16_t old_value, new_value;
+
+	rc = vcnl36825t_read(spec, reg_addr, &old_value);
+	if (rc < 0) {
+		return rc;
+	}
+
+	new_value = (old_value & ~mask) | (value & mask);
+
+	if (new_value == old_value) {
+		return 0;
+	}
+
+	return vcnl36825t_write(spec, reg_addr, new_value);
+}
+
+#if CONFIG_PM_DEVICE
+static int vcnl36825t_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	const struct vcnl36825t_config *config = dev->config;
+	int rc = 0;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_RESUME:
+		rc = vcnl36825t_update(&config->i2c, VCNL36825T_REG_PS_CONF1, VCNL36825T_PS_ON_MSK,
+				       VCNL36825T_PS_ON);
+		if (rc < 0) {
+			return rc;
+		}
+
+		rc = vcnl36825t_update(&config->i2c, VCNL36825T_REG_PS_CONF2, VCNL36825T_PS_ST_MSK,
+				       VCNL36825T_PS_ST_START);
+		if (rc < 0) {
+			return rc;
+		}
+		break;
+	case PM_DEVICE_ACTION_SUSPEND:
+		rc = vcnl36825t_update(&config->i2c, VCNL36825T_REG_PS_CONF2, VCNL36825T_PS_ST_MSK,
+				       VCNL36825T_PS_ST_STOP);
+		if (rc < 0) {
+			return rc;
+		}
+
+		rc = vcnl36825t_update(&config->i2c, VCNL36825T_REG_PS_CONF1, VCNL36825T_PS_ON_MSK,
+				       VCNL36825T_PS_OFF);
+		if (rc < 0) {
+			return rc;
+		}
```

Comment:
After a suspend/resume are all reg values preserved, correct? I'm assuming it is correct as the proximity sensor should not be powered-off.

Reply:
Registers should be preserved according to the datasheet.

Setting `VCNL36825T_PS_OFF` stops the bias current (according to the datasheet), but the sensor stays powered without performing any operation.